# iNaturalist status updates by state

Using the file produced in the collate-status-taxa.ipynb: `inat-aust-status-taxa.csv`, generate lists to update iNaturalist statuses

**Next steps:**
State by state establish the changes that need to be made:
    a. new - any new species that appear in the state lists but do not have a status in inaturalist (new template)
    b. updates - any changes to information which was added by us previously (user_id = 708886) (update template, action='UPDATE')
    c. removals - any statuses which were added by us previously (user_id = 708886) list which are incorrect (update template, action='REMOVE')
    d. flags - are there any statuses by other users that need to be flagged?

## Prep - common to all states
1. Read in the inaturalist statuses & filter out this state
2. Read in the inaturalist taxa list
3. Read in the state sensitive list
4. Attempt to match the state statuses to an IUCN equivalent


### 1. iNaturalist statuses

In [1]:
import sys
import os
projectdir = "/Users/new330/IdeaProjects/authoritative-lists/" # basedir for this gh project
#projectdir = "/Users/oco115/PycharmProjects/authoritative-lists/" # basedir for this gh project
sys.path.append(os.path.abspath(projectdir + "source-code/includes"))
import list_functions as lf

import pandas as pd
# projectdir = "/Users/new330/IdeaProjects/authoritative-lists/" # basedir for this gh project
sourcedir = projectdir + "source-data/inaturalist-statuses/"
listdir = projectdir + "current-lists/"


# read in the statuses
taxastatus = pd.read_csv(sourcedir + "inat-aust-status-taxa.csv", encoding='UTF-8',na_filter=False,dtype=str) ## Read inaturalist conservation statuses file
taxastatus.head(3)

id taxon_id  user_id place_id source_id  \
0  166449    38493  1138587     7830             
1  234788   918383   702203     9994             
2  234789   918383   702203     7308             

                            authority status  \
0  Flora and Fauna Guarantee Act 1988     CR   
1           Atlas of Living Australia    NT    
2           Atlas of Living Australia     LC   

                                                 url description geoprivacy  \
0                                                                  obscured   
1  https://bie.ala.org.au/species/https://id.biod...                          
2  https://bie.ala.org.au/species/https://id.biod...                          

   ...         genus specificEpithet infraspecificEpithet  \
0  ...     Eulamprus      kosciuskoi                        
1  ...  Chiloschista     phyllorhiza                        
2  ...  Chiloschista     phyllorhiza                        

               modified            scientificName taxonRank  \
0  2021-03-01T10:35:01Z      Eulamprus kosciuskoi   species   
1  2022-01-08T03:30:36Z  Chiloschista phyllorhiza   species   
2  2022-01-08T03:30:36Z  Chiloschista phyllorhiza   species   

                                          references preferred_common_name  \
0  http://reptile-database.reptarium.cz/search.ph...                         
1  http://www.catalogueoflife.org/annual-checklis...                         
2  http://www.catalogueoflife.org/annual-checklis...                         

  is_active current_synonymous_taxon_ids  
0                                         
1                                         
2                                         

[3 rows x 34 columns]

In [2]:
def filter_state_statuses(stateregex: str, urlregex: str):
    authoritydf = taxastatus['authority'].drop_duplicates().sort_values()
    authoritydf = authoritydf[pd.Series(authoritydf).str.contains(stateregex)]
    urldf = taxastatus['url'].drop_duplicates().sort_values()
    urldf = urldf[pd.Series(urldf).str.contains(urlregex)]
    placedisplaydf = taxastatus['place_display_name'].drop_duplicates().sort_values()
    placedisplaydf = placedisplaydf[pd.Series(placedisplaydf).str.contains(stateregex)]
    placedf = taxastatus['place_name'].drop_duplicates().sort_values()
    placedf = placedf[pd.Series(placedf).str.contains(stateregex)]
    # concat all and remove duplicates
    statedf = pd.concat([taxastatus.apply(lambda row: row[taxastatus['place_display_name'].isin(placedisplaydf)]),
                         taxastatus.apply(lambda row: row[taxastatus['place_name'].isin(placedf)]),
                         taxastatus.apply(lambda row: row[taxastatus['url'].isin(urldf)]),
                         taxastatus.apply(
                             lambda row: row[taxastatus['authority'].isin(authoritydf)])]).drop_duplicates()
    return statedf.sort_values(['taxon_id', 'user_id'])
inatstatuses = filter_state_statuses("Tasmania|TAS", ".tas.gov")
inatstatuses.rename(columns={'id':'status_id','id_y':'taxon_id_y'},inplace=True)
inatstatuses

status_id taxon_id  user_id place_id source_id  \
212     170159  1007040   702203     6829             
2558    160921  1038747   717122     6829             
2547    160910  1038906   717122     6829             
933     160738  1038964   717122     6829             
2557    160920  1038976   717122     6829             
...        ...      ...      ...      ...       ...   
2586    160759   958863   717122     6829             
2594    161103   960004   717122     6829             
3380    268029   993886  1116633     6829             
813     160748   993902   717122     6829             
1008    160745   993922   717122     6829             

                                            authority      status  \
212            Threatened Species Protection Act 1995          EN   
2558           Threatened Species Protection Act 1995  Endangered   
2547           Threatened Species Protection Act 1995  Endangered   
933            Threatened Species Protection Act 1995  Endangered   
2557           Threatened Species Protection Act 1995  Endangered   
...                                               ...         ...   
2586           Threatened Species Protection Act 1995  Endangered   
2594           Threatened Species Protection Act 1995  Endangered   
3380  TASMANIAN THREATENED SPECIES LISTING STATEMENT           EN   
813            Threatened Species Protection Act 1995  Endangered   
1008           Threatened Species Protection Act 1995  Endangered   

                                                    url  \
212   https://www.threatenedspecieslink.tas.gov.au/P...   
2558  https://dpipwe.tas.gov.au/Documents/Ranunculus...   
2547  https://dpipwe.tas.gov.au/Documents/Myosurus-a...   
933   https://dpipwe.tas.gov.au/Documents/Boronia%20...   
2557  https://dpipwe.tas.gov.au/Documents/Pimelea-ax...   
...                                                 ...   
2586  https://dpipwe.tas.gov.au/Documents/Corunastyl...   
2594  https://dpipwe.tas.gov.au/Documents/Thelymitra...   
3380  https://nre.tas.gov.au/Documents/Prasophyllum%...   
813   https://dpipwe.tas.gov.au/Documents/Caladenia-...   
1008  https://dpipwe.tas.gov.au/Documents/Caladenia-...   

                                            description geoprivacy  ...  \
212                                                                 ...   
2558                                                      obscured  ...   
2547                                                      obscured  ...   
933                                                       obscured  ...   
2557                                                      obscured  ...   
...                                                 ...        ...  ...   
2586                                                                ...   
2594                                                      obscured  ...   
3380  1. total population estimated to number fewer ...   obscured  ...   
813                                                                 ...   
1008                                                                ...   

             genus specificEpithet infraspecificEpithet              modified  \
212       Dianella          amoena                       2021-07-27T17:12:17Z   
2558    Ranunculus       diminutus                       2022-06-13T17:38:59Z   
2547      Myosurus       australis                       2020-08-19T03:33:32Z   
933        Boronia      hemichiton                       2020-08-17T05:51:50Z   
2557       Pimelea        axiflora             axiflora  2020-12-02T04:00:26Z   
...            ...             ...                  ...                   ...   
2586   Genoplesium      nudiscapum                       2022-10-26T09:34:12Z   
2594    Thelymitra       bracteata                       2022-07-14T00:09:44Z   
3380  Prasophyllum    abblittiorum                       2020-02-18T23:45:10Z   
813      Caladenia         dienema                       2020-12-15T00:50:01Z   
1008     Calad

### 2. iNaturalist taxonomy

In [3]:
# Output files contain these fields
# Taxon_Name,Status,Authority,IUCN_equivalent,Description,iNaturalist_Place_ID,url,Taxon_Geoprivacy,Username,taxon_id
# so we need to match species from the state lists to the inat taxa to get the taxon_id

import zipfile
url = "https://www.inaturalist.org/taxa/inaturalist-taxonomy.dwca.zip"
filename = url.split("/")[-1]

z=zipfile.ZipFile(sourcedir + filename)

with z.open('taxa.csv') as from_archive:
    inattaxa = pd.read_csv(from_archive,dtype=str)
z.close()
inattaxa.head(3)


id                             taxonID                          identifier  \
0  1  https://www.inaturalist.org/taxa/1  https://www.inaturalist.org/taxa/1   
1  2  https://www.inaturalist.org/taxa/2  https://www.inaturalist.org/taxa/2   
2  3  https://www.inaturalist.org/taxa/3  https://www.inaturalist.org/taxa/3   

                         parentNameUsageID   kingdom    phylum class order  \
0   https://www.inaturalist.org/taxa/48460  Animalia       NaN   NaN   NaN   
1       https://www.inaturalist.org/taxa/1  Animalia  Chordata   NaN   NaN   
2  https://www.inaturalist.org/taxa/355675  Animalia  Chordata  Aves   NaN   

  family genus specificEpithet infraspecificEpithet              modified  \
0    NaN   NaN             NaN                  NaN  2021-11-02T06:05:44Z   
1    NaN   NaN             NaN                  NaN  2021-11-23T00:40:18Z   
2    NaN   NaN             NaN                  NaN  2022-12-27T07:33:16Z   

  scientificName taxonRank                                         references  
0       Animalia   kingdom  http://www.catalogueoflife.org/annual-checklis...  
1       Chordata    phylum  http://www.catalogueoflife.org/annual-checklis...  
2           Aves     class  http://www.catalogueoflife.org/annual-checklis...

### 3. State lists
Sensitive list: `geoprivacy` = `obscured`
overrides

In [4]:
# %%script echo skipping # comment this line to download dataset from lists.ala.org.au the web and save locally
# Download lists data. Retrieve binomial and trinomial names from GBIF. Save locally to CSV

sensitivelist = lf.download_ala_list("https://lists.ala.org.au/ws/speciesListItems/dr491?max=10000&includeKVP=true")
# sensitivelist = lf.kvp_to_columns(sensitivelist)   # kvp values null
sensitivelist.to_csv(sourcedir + "tas-ala-sensitive.csv", index=False)

In [5]:
conservationlist = lf.download_ala_list("https://lists.ala.org.au/ws/speciesListItems/dr654?max=10000&includeKVP=true")
conservationlist = lf.kvp_to_columns(conservationlist)
conservationlist.to_csv(sourcedir + "tas-ala-conservation.csv", index=False)

In [19]:
# Read sensitive list data
sensitivelist = pd.read_csv(sourcedir + "tas-ala-sensitive.csv", dtype=str)
# Tasmania has only 1 sensitive species in list so set status to Conservation list status for the species
sensitivelist['status'] = 'Endangered'
sensitivelist['tas_geoprivacy'] = 'private'
sensitivelist

id               name      commonName     scientificName  \
0  24386  Lomatia tasmanica  King's Lomatia  Lomatia tasmanica   

                                               lsid dataResourceUid kvpValues  \
0  https://id.biodiversity.org.au/node/apni/2903191           dr491        []   

       status tas_geoprivacy  
0  Endangered        private

In [8]:
# Read conservation list data
conservationlist = pd.read_csv(sourcedir + "tas-ala-conservation.csv", dtype=str)
conservationlist['tas_geoprivacy'] = 'open'
conservationlist['tas_geoprivacy'] = conservationlist['name'].apply(lambda x: 'private' if x == "Lomatia tasmanica" else 'open')
conservationlist

id                                    name  \
0    4033150                           Costora iena    
1    4033128                  Hadronyche pulvinator    
2    4033378                 Hypolimnus pedderensis    
3    4033234                Thylacinus cynocephalus    
4    4033416  Gallirallus philippensis macquariensis   
..       ...                                     ...   
678  4033369                    Erioderma sorediatum   
679  4033419              Xanthoparmelia amphixantha   
680  4033291              Xanthoparmelia molliuscula   
681  4033236             Xanthoparmelia subloxodella   
682  4033313                 Xanthoparmelia willisii   

                    commonName                          scientificName  \
0     Caddis Fly (great Lakes)                            Costora iena   
1    Cascade Funnel-web Spider                   Hadronyche pulvinator   
2        Lake Pedder Earthworm                  Hypolimnus pedderensis   
3                    Thylacine                 Thylacinus cynocephalus   
4                          NaN  Gallirallus philippensis macquariensis   
..                         ...                                     ...   
678                        NaN                    Erioderma sorediatum   
679                        NaN              Xanthoparmelia amphixantha   
680                        NaN              Xanthoparmelia molliuscula   
681                        NaN             Xanthoparmelia subloxodella   
682                        NaN                 Xanthoparmelia willisii   

                                                  lsid dataResourceUid  \
0    https://biodiversity.org.au/afd/taxa/7de808fb-...           dr654   
1    https://biodiversity.org.au/afd/taxa/2acd0df1-...           dr654   
2    https://biodiversity.org.au/afd/taxa/7db09aa3-...           dr654   
3    https://biodiversity.org.au/afd/taxa/97764bed-...           dr654   
4                                        NZOR-6-101588           dr654   
..                                                 ...             ...   
678                                      ALA_DR654_679           dr654   
679                                      ALA_DR654_680           dr654   
680                                      ALA_DR654_681           dr654   
681                                      ALA_DR654_682           dr654   
682                                      ALA_DR654_683           dr654   

                                             kvpValues  \
0    [{'key': 'scientificNameAuthorship', 'value': ...   
1    [{'key': 'scientificNameAuthorship', 'value': ...   
2    [{'key': 'scientificNameAuthorship', 'value': ...   
3    [{'key': 'scientificNameAuthorship', 'value': ...   
4    [{'key': 'scientificNameAuthorship', 'value': ...   
..                                                 ...   
678  [{'key': 'scientificNameAuthorship', 'value': ...   
679  [{'key': 'scientificNameAuthorship', 'value': ...   
680  [{'key': 'scientificNameAuthorship', 'value': ...   
681  [{'key': 'scientificNameAuthorship', 'value': ...   
682  [{'key': 'scientificNameAuthorship', 'value': ...   

                              scientificNameAuthorship  \
0                                               Mosely   
1                                              Hickman   
2                                           (Jamieson)   
3                                               Harris   
4                                               Hutton   
..                                                 ...   
678                           D.J.Galloway & P.M.Jørg.   
679                                   (Müll.Arg.) Hale   
680                                        (Ach.) Hale   
681  (Elix & Kantvilas) O.Blanco, A.Crespo, Elix, D...   
682                 (Kurok. & Filson) Elix & J.Johnst.   

               vernacular name        family          class sourceStatus  \
0     Caddis Fly (Great Lakes)   Trichoptera  Invertebrates            x   
1    Cascade 

In [20]:
# join them in a way that works for inat (eg sensitive list, geoprivacy = 'obscured'
statelist = pd.concat([sensitivelist[['name','tas_geoprivacy', 'status', 'lsid']],
                    conservationlist[['name','tas_geoprivacy', 'status', 'lsid']]]).drop_duplicates()

In [11]:
# retrieve binomial and trinomial names from GBIF
parsednames = lf.gbifparse(statelist)
parsednames.to_csv(sourcedir + "tas-gbif.csv", index=False)

In [21]:
parsednames = pd.read_csv(sourcedir + "tas-gbif.csv")
statelist = statelist.merge(parsednames[['scientificName','canonicalName','canonicalNameComplete','type','rankMarker']],how="left",left_on="name",right_on="scientificName")
numfullstatelist = len(statelist.index)
statelist['scientificName'] = statelist['canonicalName']
statelist

name tas_geoprivacy            status  \
0              Lomatia tasmanica        private        Endangered   
1              Lomatia tasmanica        private        Endangered   
2                  Costora iena            open  Presumed Extinct   
3         Hadronyche pulvinator            open  Presumed Extinct   
4        Hypolimnus pedderensis            open  Presumed Extinct   
..                           ...            ...               ...   
679         Erioderma sorediatum           open        Endangered   
680   Xanthoparmelia amphixantha           open        Endangered   
681   Xanthoparmelia molliuscula           open        Endangered   
682  Xanthoparmelia subloxodella           open        Endangered   
683      Xanthoparmelia willisii           open        Endangered   

                                                  lsid  \
0     https://id.biodiversity.org.au/node/apni/2903191   
1     https://id.biodiversity.org.au/node/apni/2903191   
2    https://biodiversity.org.au/afd/taxa/7de808fb-...   
3    https://biodiversity.org.au/afd/taxa/2acd0df1-...   
4    https://biodiversity.org.au/afd/taxa/7db09aa3-...   
..                                                 ...   
679                                      ALA_DR654_679   
680                                      ALA_DR654_680   
681                                      ALA_DR654_681   
682                                      ALA_DR654_682   
683                                      ALA_DR654_683   

                  scientificName                canonicalName  \
0              Lomatia tasmanica            Lomatia tasmanica   
1              Lomatia tasmanica            Lomatia tasmanica   
2                   Costora iena                 Costora iena   
3          Hadronyche pulvinator        Hadronyche pulvinator   
4         Hypolimnus pedderensis       Hypolimnus pedderensis   
..                           ...                          ...   
679         Erioderma sorediatum         Erioderma sorediatum   
680   Xanthoparmelia amphixantha   Xanthoparmelia amphixantha   
681   Xanthoparmelia molliuscula   Xanthoparmelia molliuscula   
682  Xanthoparmelia subloxodella  Xanthoparmelia subloxodella   
683      Xanthoparmelia willisii      Xanthoparmelia willisii   

           canonicalNameComplete        type rankMarker  
0              Lomatia tasmanica  SCIENTIFIC        sp.  
1              Lomatia tasmanica  SCIENTIFIC        sp.  
2                   Costora iena  SCIENTIFIC        sp.  
3          Hadronyche pulvinator  SCIENTIFIC        sp.  
4         Hypolimnus pedderensis  SCIENTIFIC        sp.  
..                           ...         ...        ...  
679         Erioderma sorediatum  SCIENTIFIC        sp.  
680   Xanthoparmelia amphixantha  SCIENTIFIC        sp.  
681   Xanthoparmelia molliuscula  SCIENTIFIC        sp.  
682  Xanthoparmelia subloxodella  SCIENTIFIC        sp.  
683      Xanthoparmelia willisii  SCIENTIFIC        sp.  

[684 rows x 9 columns]

In [22]:
# Identify records that won't comply with iNaturalist species names
noncomply = statelist[statelist['type'].isin(['INFORMAL','CULTIVAR','HYBRID']) ]
noncomply

name tas_geoprivacy      status           lsid  \
79             Charopidae "Skemps"           open        Rare   ALA_DR654_78   
83              Enchymus sp. nov.            open        Rare   ALA_DR654_82   
324      Isoetes sp. Maxwell River           open        Rare  ALA_DR654_323   
569      Euphrasia sp. Bivouac Bay           open  Endangered  ALA_DR654_568   
587  Prasophyllum sp. Arthurs Lake           open  Endangered  ALA_DR654_587   
647          Pimelea sp. Tunbridge           open  Endangered  ALA_DR654_647   
649  Polyscias sp. Douglas-Denison           open  Endangered  ALA_DR654_649   

                   scientificName                 canonicalName  \
79                 Charopidae cv.                Charopidae cv.   
83                 Enchymus spec.                Enchymus spec.   
324            Isoetes sp.Maxwell            Isoetes sp.Maxwell   
569          Euphrasia sp.Bivouac          Euphrasia sp.Bivouac   
587       Prasophyllum sp.Arthurs       Prasophyllum sp.Arthurs   
647          Pimelea sp.Tunbridge          Pimelea sp.Tunbridge   
649  Polyscias sp.Douglas-Denison  Polyscias sp.Douglas-Denison   

            canonicalNameComplete      type rankMarker  
79        Charopidae cv. 'Skemps'  CULTIVAR        cv.  
83                 Enchymus spec.  INFORMAL        sp.  
324            Isoetes sp.Maxwell  INFORMAL        sp.  
569          Euphrasia sp.Bivouac  INFORMAL        sp.  
587       Prasophyllum sp.Arthurs  INFORMAL        sp.  
647          Pimelea sp.Tunbridge  INFORMAL        sp.  
649  Polyscias sp.Douglas-Denison  INFORMAL        sp.

In [23]:
# remove records that do not comply
statelist = statelist[~statelist['type'].isin(['INFORMAL','CULTIVAR','HYBRID']) ]
statelist = pd.DataFrame(statelist[['scientificName','status','tas_geoprivacy','lsid']]).drop_duplicates()
statelist

scientificName            status tas_geoprivacy  \
0              Lomatia tasmanica        Endangered        private   
2                   Costora iena  Presumed Extinct           open   
3          Hadronyche pulvinator  Presumed Extinct           open   
4         Hypolimnus pedderensis  Presumed Extinct           open   
5        Thylacinus cynocephalus  Presumed Extinct           open   
..                           ...               ...            ...   
679         Erioderma sorediatum        Endangered           open   
680   Xanthoparmelia amphixantha        Endangered           open   
681   Xanthoparmelia molliuscula        Endangered           open   
682  Xanthoparmelia subloxodella        Endangered           open   
683      Xanthoparmelia willisii        Endangered           open   

                                                  lsid  
0     https://id.biodiversity.org.au/node/apni/2903191  
2    https://biodiversity.org.au/afd/taxa/7de808fb-...  
3    https://biodiversity.org.au/afd/taxa/2acd0df1-...  
4    https://biodiversity.org.au/afd/taxa/7db09aa3-...  
5    https://biodiversity.org.au/afd/taxa/97764bed-...  
..                                                 ...  
679                                      ALA_DR654_679  
680                                      ALA_DR654_680  
681                                      ALA_DR654_681  
682                                      ALA_DR654_682  
683                                      ALA_DR654_683  

[676 rows x 4 columns]

In [16]:
parsednames['type'].unique()

array(['SCIENTIFIC', 'CULTIVAR', 'INFORMAL'], dtype=object)

In [24]:
## Testing to update duplicates
statelist

scientificName            status tas_geoprivacy  \
0              Lomatia tasmanica        Endangered        private   
2                   Costora iena  Presumed Extinct           open   
3          Hadronyche pulvinator  Presumed Extinct           open   
4         Hypolimnus pedderensis  Presumed Extinct           open   
5        Thylacinus cynocephalus  Presumed Extinct           open   
..                           ...               ...            ...   
679         Erioderma sorediatum        Endangered           open   
680   Xanthoparmelia amphixantha        Endangered           open   
681   Xanthoparmelia molliuscula        Endangered           open   
682  Xanthoparmelia subloxodella        Endangered           open   
683      Xanthoparmelia willisii        Endangered           open   

                                                  lsid  
0     https://id.biodiversity.org.au/node/apni/2903191  
2    https://biodiversity.org.au/afd/taxa/7de808fb-...  
3    https://biodiversity.org.au/afd/taxa/2acd0df1-...  
4    https://biodiversity.org.au/afd/taxa/7db09aa3-...  
5    https://biodiversity.org.au/afd/taxa/97764bed-...  
..                                                 ...  
679                                      ALA_DR654_679  
680                                      ALA_DR654_680  
681                                      ALA_DR654_681  
682                                      ALA_DR654_682  
683                                      ALA_DR654_683  

[676 rows x 4 columns]

In [25]:
# check for duplicates with conflicting information
dupinformation = statelist.groupby('scientificName').filter(lambda x: len(x) > 1)#.sort('size',ascending=False)
dupinformation

Empty DataFrame
Columns: [scientificName, status, tas_geoprivacy, lsid]
Index: []

### 4. Equivalent IUCN statuses

In [26]:
# iucn_statuses = {'Not Evaluated', 'Data Deficient', 'Least Concern', 'Near Threatened', 'Vulnerable', 'Endangered', 'Critically Endangered', 'Extinct in the Wild', 'Extinct'}
statelist.groupby(['status'])['status'].count()

status
Endangered          218
Presumed Extinct     28
Rare                308
Vulnerable          122
Name: status, dtype: int64

In [27]:
iucnStatusMappings = {
    'presumed extinct':'Extinct',
    'rare':'Vulnerable',
    'sensitive':'Vulnerable',
    'endangered': 'Endangered',
    'vulnerable': 'Vulnerable'
}

### 5. Determine best place ID to use

In [28]:
inatstatuses.groupby(['place_id','place_name','place_display_name'])['place_id'].count()
# looks like 6829

place_id  place_name          place_display_name    
6744      Australia           Australia                  2
6829      Tasmania            Tasmania, AU              77
9994      Northern Territory  Northern Territory, AU     6
Name: place_id, dtype: int64

## Merge iNaturalist statuses with State sensitive list on scientificName

1. Match - updates, even if the statuses are the same we'll update the links and values anyway
2. No match - statuses to be added (additions)
   1.1 No match and no taxnomy - search for synonyms
   1.2 No match
3. Merge the other direction to see if there are deletes?


In [40]:
# join to see which lists already have a status in inaturalist based on scientificName
mergedstatuses = statelist[['scientificName', 'status','tas_geoprivacy', 'lsid']].merge(inatstatuses[['status_id','scientificName','taxon_id','user_id','description','iucn','authority','status','geoprivacy','place_id','place_display_name']],how="left",left_on='scientificName',right_on='scientificName',suffixes=(None,'_inat')).sort_values(['scientificName'])
mergedstatuses


scientificName      status tas_geoprivacy  \
214            Acacia axillaris  Vulnerable           open   
298           Acacia pataczekii        Rare           open   
352         Acacia siculiformis        Rare           open   
353           Acacia ulicifolia        Rare           open   
354            Acacia uncifolia        Rare           open   
..                          ...         ...            ...   
285        Xerochrysum palustre  Vulnerable           open   
171            Zearaja maugeana  Endangered           open   
519           Zieria littoralis        Rare           open   
670  Zieria veronicea veronicea  Endangered           open   
520    Zygophyllum billardierei        Rare           open   

                                                  lsid status_id taxon_id  \
214   https://id.biodiversity.org.au/node/apni/2911902    160726   775192   
298   https://id.biodiversity.org.au/node/apni/2913609       NaN      NaN   
352   https://id.biodiversity.org.au/node/apni/2888960       NaN      NaN   
353   https://id.biodiversity.org.au/node/apni/2907350       NaN      NaN   
354   https://id.biodiversity.org.au/node/apni/2916270       NaN      NaN   
..                                                 ...       ...      ...   
285   https://id.biodiversity.org.au/node/apni/2890165       NaN      NaN   
171  https://biodiversity.org.au/afd/taxa/0d9d194f-...       NaN      NaN   
519   https://id.biodiversity.org.au/node/apni/2888662       NaN      NaN   
670   https://id.biodiversity.org.au/node/apni/2911634    161108   323791   
520  https://id.biodiversity.org.au/taxon/apni/5131...       NaN      NaN   

    user_id description iucn  \
214  717122               30   
298     NaN         NaN  NaN   
352     NaN         NaN  NaN   
353     NaN         NaN  NaN   
354     NaN         NaN  NaN   
..      ...         ...  ...   
285     NaN         NaN  NaN   
171     NaN         NaN  NaN   
519     NaN         NaN  NaN   
670  717122               40   
520     NaN         NaN  NaN   

                                             authority status_inat geoprivacy  \
214  Environment Protection and Biodiversity Conser...          VU   obscured   
298                                                NaN         NaN        NaN   
352                                                NaN         NaN        NaN   
353                                                NaN         NaN        NaN   
354                                                NaN         NaN        NaN   
..                                                 ...         ...        ...   
285                                                NaN         NaN        NaN   
171                                                NaN         NaN        NaN   
519                                                NaN         NaN        NaN   
670             Threatened Species Protection Act 1995  Endangered   obscured   
520                                                NaN         NaN        NaN   

    place_id place_display_name  
214     6744          Australia  
298      NaN                NaN  
352      NaN                NaN  
353      NaN                NaN  
354      NaN                NaN  
..       ...                ...  
285      NaN                NaN  
171      NaN                NaN  
519      NaN                NaN  
670     6829       Tasmania, AU  
520      NaN                NaN  

[676 rows x 14 columns]

In [41]:
# prepare the export fields, common to New template and Update template
# new statuses
# Taxon_Name,Status,Authority,IUCN_equivalent,Description,iNaturalist_Place_ID,url,Taxon_Geoprivacy,Username,taxon_id
# updates
# action,taxon_name,id,taxon_id,status,iucn_equivalent,authority,url,geoprivacy,place_id,username
mergedstatuses['new_authority'] = "Threatened Species Protection Act 1995"
mergedstatuses['new_description'] = "Listed as Threatened - refer to https://www.threatenedspecieslink.tas.gov.au"
biesearchurl = "https://bie.ala.org.au/species/" # eg + "https://id.biodiversity.org.au/node/apni/2894366"
mergedstatuses['new_url'] =  biesearchurl + mergedstatuses['lsid']
mergedstatuses['new_geoprivacy'] = "obscured"
mergedstatuses['new_place_id'] = '6829'  # Tasmania
mergedstatuses['new_username'] = 'peggydnew'
mergedstatuses['new_iucn_equivalent'] = mergedstatuses['status'].str.lower().str.strip().map(iucnStatusMappings).fillna('Vulnerable') # map to dictionary
mergedstatuses['new_status'] = mergedstatuses['status'].fillna('Threatened')
mergedstatuses

scientificName      status tas_geoprivacy  \
214            Acacia axillaris  Vulnerable           open   
298           Acacia pataczekii        Rare           open   
352         Acacia siculiformis        Rare           open   
353           Acacia ulicifolia        Rare           open   
354            Acacia uncifolia        Rare           open   
..                          ...         ...            ...   
285        Xerochrysum palustre  Vulnerable           open   
171            Zearaja maugeana  Endangered           open   
519           Zieria littoralis        Rare           open   
670  Zieria veronicea veronicea  Endangered           open   
520    Zygophyllum billardierei        Rare           open   

                                                  lsid status_id taxon_id  \
214   https://id.biodiversity.org.au/node/apni/2911902    160726   775192   
298   https://id.biodiversity.org.au/node/apni/2913609       NaN      NaN   
352   https://id.biodiversity.org.au/node/apni/2888960       NaN      NaN   
353   https://id.biodiversity.org.au/node/apni/2907350       NaN      NaN   
354   https://id.biodiversity.org.au/node/apni/2916270       NaN      NaN   
..                                                 ...       ...      ...   
285   https://id.biodiversity.org.au/node/apni/2890165       NaN      NaN   
171  https://biodiversity.org.au/afd/taxa/0d9d194f-...       NaN      NaN   
519   https://id.biodiversity.org.au/node/apni/2888662       NaN      NaN   
670   https://id.biodiversity.org.au/node/apni/2911634    161108   323791   
520  https://id.biodiversity.org.au/taxon/apni/5131...       NaN      NaN   

    user_id description iucn  \
214  717122               30   
298     NaN         NaN  NaN   
352     NaN         NaN  NaN   
353     NaN         NaN  NaN   
354     NaN         NaN  NaN   
..      ...         ...  ...   
285     NaN         NaN  NaN   
171     NaN         NaN  NaN   
519     NaN         NaN  NaN   
670  717122               40   
520     NaN         NaN  NaN   

                                             authority  ... place_id  \
214  Environment Protection and Biodiversity Conser...  ...     6744   
298                                                NaN  ...      NaN   
352                                                NaN  ...      NaN   
353                                                NaN  ...      NaN   
354                                                NaN  ...      NaN   
..                                                 ...  ...      ...   
285                                                NaN  ...      NaN   
171                                                NaN  ...      NaN   
519                                                NaN  ...      NaN   
670             Threatened Species Protection Act 1995  ...     6829   
520                                                NaN  ...      NaN   

    place_display_name                           new_authority  \
214          Australia  Threatened Species Protection Act 1995   
298                NaN  Threatened Species Protection Act 1995   
352                NaN  Threatened Species Protection Act 1995   
353                NaN  Threatened Species Protection Act 1995   
354                NaN  Threatened Species Protection Act 1995   
..                 ...                                     ...   
285                NaN  Threatened Species Protection Act 1995   
171                NaN  Threatened Species Protection Act 1995   
519                NaN  Threatened Species Protection Act 1995   
670       Tasmania, AU  Threatened Species Protection Act 1995   
520                NaN  Threatened Species Protection Act 1995   

                                       new_description  \
214  Listed as Threatened - refer to https://www.th...   
298  Listed as Threatened - refer to https://www.th...   
352  Listed as Threatened - refer to https://www.th...   
353  Listed as Threatened - refer to https://www.th...   
354  Listed as

## Updates

In [42]:
# updates
# action,taxon_name,id,taxon_id,status,iucn_equivalent,authority,url,geoprivacy,place_id,username,description
updates = pd.DataFrame(mergedstatuses[mergedstatuses['status_id'].notnull()])
updates.sort_values('scientificName')
updates['action'] = 'UPDATE'
#updates.loc[:,'action'] = 'UPDATE'
updates = updates[['action','scientificName','status_id','taxon_id','new_status','new_iucn_equivalent','new_authority','new_url','new_geoprivacy','new_place_id','new_username','new_description']]
updates.columns = updates.columns.str.replace("new_", "", regex=True)
updates = updates.rename(columns={'scientificName':'taxon_name',
                                  'status_id':'id'})
updates

action                  taxon_name      id taxon_id      status  \
214  UPDATE            Acacia axillaris  160726   775192  Vulnerable   
598  UPDATE   Alternanthera denticulata  160727   369665  Endangered   
599  UPDATE     Amphibromus macrorhinus  160728  1121410  Endangered   
241  UPDATE       Anogramma leptophylla  162081   180646  Vulnerable   
357  UPDATE            Aphelia gracilis  163877  1143399        Rare   
..      ...                         ...     ...      ...         ...   
663  UPDATE          Thelymitra malvina  161104   406524  Endangered   
666  UPDATE          Veronica notabilis  161105   791936  Endangered   
667  UPDATE            Viminaria juncea  161106   353911  Endangered   
669  UPDATE  Wurmbea latifolia vanessae  161107   790432  Endangered   
670  UPDATE  Zieria veronicea veronicea  161108   323791  Endangered   

    iucn_equivalent                               authority  \
214      Vulnerable  Threatened Species Protection Act 1995   
598      Endangered  Threatened Species Protection Act 1995   
599      Endangered  Threatened Species Protection Act 1995   
241      Vulnerable  Threatened Species Protection Act 1995   
357      Vulnerable  Threatened Species Protection Act 1995   
..              ...                                     ...   
663      Endangered  Threatened Species Protection Act 1995   
666      Endangered  Threatened Species Protection Act 1995   
667      Endangered  Threatened Species Protection Act 1995   
669      Endangered  Threatened Species Protection Act 1995   
670      Endangered  Threatened Species Protection Act 1995   

                                                   url geoprivacy place_id  \
214  https://bie.ala.org.au/species/https://id.biod...   obscured     6829   
598  https://bie.ala.org.au/species/https://id.biod...   obscured     6829   
599  https://bie.ala.org.au/species/https://id.biod...   obscured     6829   
241  https://bie.ala.org.au/species/https://id.biod...   obscured     6829   
357  https://bie.ala.org.au/species/https://id.biod...   obscured     6829   
..                                                 ...        ...      ...   
663  https://bie.ala.org.au/species/https://id.biod...   obscured     6829   
666  https://bie.ala.org.au/species/https://id.biod...   obscured     6829   
667  https://bie.ala.org.au/species/https://id.biod...   obscured     6829   
669  https://bie.ala.org.au/species/https://id.biod...   obscured     6829   
670  https://bie.ala.org.au/species/https://id.biod...   obscured     6829   

      username                                        description  
214  peggydnew  Listed as Threatened - refer to https://www.th...  
598  peggydnew  Listed as Threatened - refer to https://www.th...  
599  peggydnew  Listed as Threatened - refer to https://www.th...  
241  peggydnew  Listed as Threatened - refer to https://www.th...  
357  peggydnew  Listed as Threatened - refer to https://www.th...  
..         ...                                                ...  
663  peggydnew  Listed as Threatened - refer to https://www.th...  
666  peggydnew  Listed as Threatened - refer to https://www.th...  
667  peggydnew  Listed as Threatened - refer to https://www.th...  
669  peggydnew  Listed as Threatened - refer to https://www.th...  
670  peggydnew  Listed as Threatened - refer to https://www.th...  

[73 rows x 12 columns]

## No status in iNaturalist via straight scientificName match
The ACT records that didn't match up to a status in iNaturalist

In [43]:
# to add: those that have no inaturalist status - 532!!
noinatstatus = mergedstatuses[mergedstatuses['status_id'].isnull()]
# try to match the taxon name to something in inaturalist
noinatstatus = noinatstatus.merge(inattaxa, how="left", left_on="scientificName",right_on="scientificName")
noinatstatus

scientificName      status tas_geoprivacy  \
0               Acacia pataczekii        Rare           open   
1             Acacia siculiformis        Rare           open   
2               Acacia ulicifolia        Rare           open   
3                Acacia uncifolia        Rare           open   
4    Acanthiza pusilla archibaldi  Endangered           open   
..                            ...         ...            ...   
598           Xerochrysum bicolor        Rare           open   
599          Xerochrysum palustre  Vulnerable           open   
600              Zearaja maugeana  Endangered           open   
601             Zieria littoralis        Rare           open   
602      Zygophyllum billardierei        Rare           open   

                                                  lsid status_id taxon_id  \
0     https://id.biodiversity.org.au/node/apni/2913609       NaN      NaN   
1     https://id.biodiversity.org.au/node/apni/2888960       NaN      NaN   
2     https://id.biodiversity.org.au/node/apni/2907350       NaN      NaN   
3     https://id.biodiversity.org.au/node/apni/2916270       NaN      NaN   
4                                        ALA_DR654_159       NaN      NaN   
..                                                 ...       ...      ...   
598   https://id.biodiversity.org.au/node/apni/2902500       NaN      NaN   
599   https://id.biodiversity.org.au/node/apni/2890165       NaN      NaN   
600  https://biodiversity.org.au/afd/taxa/0d9d194f-...       NaN      NaN   
601   https://id.biodiversity.org.au/node/apni/2888662       NaN      NaN   
602  https://id.biodiversity.org.au/taxon/apni/5131...       NaN      NaN   

    user_id description iucn authority  ...        phylum          class  \
0       NaN         NaN  NaN       NaN  ...  Tracheophyta  Magnoliopsida   
1       NaN         NaN  NaN       NaN  ...  Tracheophyta  Magnoliopsida   
2       NaN         NaN  NaN       NaN  ...  Tracheophyta  Magnoliopsida   
3       NaN         NaN  NaN       NaN  ...  Tracheophyta  Magnoliopsida   
4       NaN         NaN  NaN       NaN  ...      Chordata           Aves   
..      ...         ...  ...       ...  ...           ...            ...   
598     NaN         NaN  NaN       NaN  ...  Tracheophyta  Magnoliopsida   
599     NaN         NaN  NaN       NaN  ...  Tracheophyta  Magnoliopsida   
600     NaN         NaN  NaN       NaN  ...           NaN            NaN   
601     NaN         NaN  NaN       NaN  ...  Tracheophyta  Magnoliopsida   
602     NaN         NaN  NaN       NaN  ...           NaN            NaN   

             order        family        genus specificEpithet  \
0          Fabales      Fabaceae       Acacia      pataczekii   
1          Fabales      Fabaceae       Acacia    siculiformis   
2          Fabales      Fabaceae       Acacia      ulicifolia   
3          Fabales      Fabaceae       Acacia       uncifolia   
4    Passeriformes  Acanthizidae    Acanthiza         pusilla   
..             ...           ...          ...             ...   
598      Asterales    Asteraceae  Xerochrysum         bicolor   
599      Asterales    Asteraceae  Xerochrysum        palustre   
600            NaN           NaN          NaN             NaN   
601     Sapindales      Rutaceae       Zieria      littoralis   
602            NaN           NaN          NaN             NaN   

    infraspecificEpithet              modified   taxonRank  \
0                    NaN  2022-04-06T23:48:02Z     species   
1                    NaN  2022-04-07T00:40:40Z     species   
2                    NaN  2022-06-18T06:33:18Z     species   
3                    NaN  2022-06-10T10:52:23Z     species   
4             archibaldi  2018-12-19T06:16:34Z  subspecies   
..                   ...                   ...         ...   
598                  NaN  2020-02-19T08:20:14Z     species   
599                  NaN  2022-06-08T17:48:57Z     species   
600                  NaN                   NaN         NaN   
601        

In [44]:
additions = pd.DataFrame(noinatstatus[noinatstatus['id'].notna()])
additions

scientificName      status tas_geoprivacy  \
0               Acacia pataczekii        Rare           open   
1             Acacia siculiformis        Rare           open   
2               Acacia ulicifolia        Rare           open   
3                Acacia uncifolia        Rare           open   
4    Acanthiza pusilla archibaldi  Endangered           open   
..                            ...         ...            ...   
596         Xanthorrhoea arenaria  Vulnerable           open   
597        Xanthorrhoea bracteata  Vulnerable           open   
598           Xerochrysum bicolor        Rare           open   
599          Xerochrysum palustre  Vulnerable           open   
601             Zieria littoralis        Rare           open   

                                                 lsid status_id taxon_id  \
0    https://id.biodiversity.org.au/node/apni/2913609       NaN      NaN   
1    https://id.biodiversity.org.au/node/apni/2888960       NaN      NaN   
2    https://id.biodiversity.org.au/node/apni/2907350       NaN      NaN   
3    https://id.biodiversity.org.au/node/apni/2916270       NaN      NaN   
4                                       ALA_DR654_159       NaN      NaN   
..                                                ...       ...      ...   
596  https://id.biodiversity.org.au/node/apni/2918552       NaN      NaN   
597  https://id.biodiversity.org.au/node/apni/2911139       NaN      NaN   
598  https://id.biodiversity.org.au/node/apni/2902500       NaN      NaN   
599  https://id.biodiversity.org.au/node/apni/2890165       NaN      NaN   
601  https://id.biodiversity.org.au/node/apni/2888662       NaN      NaN   

    user_id description iucn authority  ...        phylum          class  \
0       NaN         NaN  NaN       NaN  ...  Tracheophyta  Magnoliopsida   
1       NaN         NaN  NaN       NaN  ...  Tracheophyta  Magnoliopsida   
2       NaN         NaN  NaN       NaN  ...  Tracheophyta  Magnoliopsida   
3       NaN         NaN  NaN       NaN  ...  Tracheophyta  Magnoliopsida   
4       NaN         NaN  NaN       NaN  ...      Chordata           Aves   
..      ...         ...  ...       ...  ...           ...            ...   
596     NaN         NaN  NaN       NaN  ...  Tracheophyta     Liliopsida   
597     NaN         NaN  NaN       NaN  ...  Tracheophyta     Liliopsida   
598     NaN         NaN  NaN       NaN  ...  Tracheophyta  Magnoliopsida   
599     NaN         NaN  NaN       NaN  ...  Tracheophyta  Magnoliopsida   
601     NaN         NaN  NaN       NaN  ...  Tracheophyta  Magnoliopsida   

             order         family         genus specificEpithet  \
0          Fabales       Fabaceae        Acacia      pataczekii   
1          Fabales       Fabaceae        Acacia    siculiformis   
2          Fabales       Fabaceae        Acacia      ulicifolia   
3          Fabales       Fabaceae        Acacia       uncifolia   
4    Passeriformes   Acanthizidae     Acanthiza         pusilla   
..             ...            ...           ...             ...   
596    Asparagales  Asphodelaceae  Xanthorrhoea        arenaria   
597    Asparagales  Asphodelaceae  Xanthorrhoea       bracteata   
598      Asterales     Asteraceae   Xerochrysum         bicolor   
599      Asterales     Asteraceae   Xerochrysum        palustre   
601     Sapindales       Rutaceae        Zieria      littoralis   

    infraspecificEpithet              modified   taxonRank  \
0                    NaN  2022-04-06T23:48:02Z     species   
1                    NaN  2022-04-07T00:40:40Z     species   
2                    NaN  2022-06-18T06:33:18Z     species   
3                    NaN  2022-06-10T10:52:23Z     species   
4             archibaldi  2018-12-19T06:16:34Z  subspecies   
..                   ...                   ...         ...   
596                  NaN  2019-02-15T07:38:04Z     species   
597                  NaN  2022-08-25T07:03:45Z     species   
598                  NaN  2020-02-19T08:20:14Z     species   

In [45]:
# there's no status but there is a matching inat taxon (id is the taxon id)
additions = pd.DataFrame(noinatstatus[noinatstatus['id'].notna()])
additions.sort_values(['scientificName'])
additions['action'] = 'ADD'
additions = additions[['action','scientificName','status_id','id','new_status','new_iucn_equivalent','new_authority','new_url','new_geoprivacy','new_place_id','new_username','new_description']]
additions.columns = additions.columns.str.replace("new_", "", regex=True)
additions = additions.rename(columns={'scientificName':'taxon_name',
                                      'id':'taxon_id',
                                  'status_id':'id'})
additions

action                    taxon_name   id taxon_id      status  \
0      ADD             Acacia pataczekii  NaN  1254534        Rare   
1      ADD           Acacia siculiformis  NaN   567724        Rare   
2      ADD             Acacia ulicifolia  NaN   139880        Rare   
3      ADD              Acacia uncifolia  NaN   560392        Rare   
4      ADD  Acanthiza pusilla archibaldi  NaN   629326  Endangered   
..     ...                           ...  ...      ...         ...   
596    ADD         Xanthorrhoea arenaria  NaN   125126  Vulnerable   
597    ADD        Xanthorrhoea bracteata  NaN   895045  Vulnerable   
598    ADD           Xerochrysum bicolor  NaN   147469        Rare   
599    ADD          Xerochrysum palustre  NaN   634811  Vulnerable   
601    ADD             Zieria littoralis  NaN   896657        Rare   

    iucn_equivalent                               authority  \
0        Vulnerable  Threatened Species Protection Act 1995   
1        Vulnerable  Threatened Species Protection Act 1995   
2        Vulnerable  Threatened Species Protection Act 1995   
3        Vulnerable  Threatened Species Protection Act 1995   
4        Endangered  Threatened Species Protection Act 1995   
..              ...                                     ...   
596      Vulnerable  Threatened Species Protection Act 1995   
597      Vulnerable  Threatened Species Protection Act 1995   
598      Vulnerable  Threatened Species Protection Act 1995   
599      Vulnerable  Threatened Species Protection Act 1995   
601      Vulnerable  Threatened Species Protection Act 1995   

                                                   url geoprivacy place_id  \
0    https://bie.ala.org.au/species/https://id.biod...   obscured     6829   
1    https://bie.ala.org.au/species/https://id.biod...   obscured     6829   
2    https://bie.ala.org.au/species/https://id.biod...   obscured     6829   
3    https://bie.ala.org.au/species/https://id.biod...   obscured     6829   
4         https://bie.ala.org.au/species/ALA_DR654_159   obscured     6829   
..                                                 ...        ...      ...   
596  https://bie.ala.org.au/species/https://id.biod...   obscured     6829   
597  https://bie.ala.org.au/species/https://id.biod...   obscured     6829   
598  https://bie.ala.org.au/species/https://id.biod...   obscured     6829   
599  https://bie.ala.org.au/species/https://id.biod...   obscured     6829   
601  https://bie.ala.org.au/species/https://id.biod...   obscured     6829   

      username                                        description  
0    peggydnew  Listed as Threatened - refer to https://www.th...  
1    peggydnew  Listed as Threatened - refer to https://www.th...  
2    peggydnew  Listed as Threatened - refer to https://www.th...  
3    peggydnew  Listed as Threatened - refer to https://www.th...  
4    peggydnew  Listed as Threatened - refer to https://www.th...  
..         ...                                                ...  
596  peggydnew  Listed as Threatened - refer to https://www.th...  
597  peggydnew  Listed as Threatened - refer to https://www.th...  
598  peggydnew  Listed as Threatened - refer to https://www.th...  
599  peggydnew  Listed as Threatened - refer to https://www.th...  
601  peggydnew  Listed as Threatened - refer to https://www.th...  

[500 rows x 12 columns]

In [46]:
# write these to the file
pd.concat([updates,additions]).to_csv(sourcedir + "tas.csv", index=False)

In [47]:
# what didnt match to a taxon?
unknownToInat = noinatstatus[noinatstatus['id'].isna()]
unknownToInat

scientificName      status tas_geoprivacy  \
8            Agrostis australiensis        Rare           open   
9                Agrostis diemenica        Rare           open   
34    Austromelanelixia piliferella  Vulnerable           open   
35          Austroparmelina pallida  Endangered           open   
36         Austroparmelina whinrayi        Rare           open   
..                              ...         ...            ...   
589  Xanthoparmelia microphyllizans        Rare           open   
593          Xanthoparmelia vicaria        Rare           open   
594       Xanthoparmelia vicariella        Rare           open   
600                Zearaja maugeana  Endangered           open   
602        Zygophyllum billardierei        Rare           open   

                                                  lsid status_id taxon_id  \
8     https://id.biodiversity.org.au/node/apni/2907094       NaN      NaN   
9     https://id.biodiversity.org.au/node/apni/2905826       NaN      NaN   
34                                       ALA_DR654_288       NaN      NaN   
35                                       ALA_DR654_600       NaN      NaN   
36                                       ALA_DR654_351       NaN      NaN   
..                                                 ...       ...      ...   
589                                      ALA_DR654_530       NaN      NaN   
593                                      ALA_DR654_353       NaN      NaN   
594                                      ALA_DR654_354       NaN      NaN   
600  https://biodiversity.org.au/afd/taxa/0d9d194f-...       NaN      NaN   
602  https://id.biodiversity.org.au/taxon/apni/5131...       NaN      NaN   

    user_id description iucn authority  ... phylum class order family genus  \
8       NaN         NaN  NaN       NaN  ...    NaN   NaN   NaN    NaN   NaN   
9       NaN         NaN  NaN       NaN  ...    NaN   NaN   NaN    NaN   NaN   
34      NaN         NaN  NaN       NaN  ...    NaN   NaN   NaN    NaN   NaN   
35      NaN         NaN  NaN       NaN  ...    NaN   NaN   NaN    NaN   NaN   
36      NaN         NaN  NaN       NaN  ...    NaN   NaN   NaN    NaN   NaN   
..      ...         ...  ...       ...  ...    ...   ...   ...    ...   ...   
589     NaN         NaN  NaN       NaN  ...    NaN   NaN   NaN    NaN   NaN   
593     NaN         NaN  NaN       NaN  ...    NaN   NaN   NaN    NaN   NaN   
594     NaN         NaN  NaN       NaN  ...    NaN   NaN   NaN    NaN   NaN   
600     NaN         NaN  NaN       NaN  ...    NaN   NaN   NaN    NaN   NaN   
602     NaN         NaN  NaN       NaN  ...    NaN   NaN   NaN    NaN   NaN   

    specificEpithet infraspecificEpithet modified taxonRank references  
8               NaN                  NaN      NaN       NaN        NaN  
9               NaN                  NaN      NaN       NaN        NaN  
34              NaN                  NaN      NaN       NaN        NaN  
35              NaN                  NaN      NaN       NaN        NaN  
36              NaN                  NaN      NaN       NaN        NaN  
..              ...                  ...      ...       ...        ...  
589             NaN                  NaN      NaN       NaN        NaN  
593             NaN                  NaN      NaN       NaN        NaN  
594             NaN                  NaN      NaN       NaN        NaN  
600             NaN                  NaN      NaN       NaN        NaN  
602             NaN                  NaN      NaN       NaN        NaN  

[103 rows x 37 columns]

### are there any that need to be removed?
TAS sensitive list count: 1
TAS conservation list count: 683
TAS inat statuses count: 0

updates to inat status: 67
additional inat status: 0
TAS statuses we can't find a taxon match for in iNaturalist: 617
total: (explainable via the various genus/section entries that we matched to in the taxonomy)

inat statuses left over:  that may need checking

In [48]:
# inat statuses that aren't in added or updated
notaddedupdated = inatstatuses[~inatstatuses['taxon_id'].isin(updates['taxon_id'])]
notaddedupdated

status_id taxon_id user_id place_id source_id  \
3353    267438  1413424  717122     6829             
1251    152435    19250  708886     9994     16652   
904     160736   353835  717122     6829             
1249    152433    38633  708886     9994     16652   
932     160737   401378  717122     6829             
959     160741   405825  717122     6829             
1252    152436    40743  708886     9994     16652   
1248    152432    41326  708886     9994     16652   
2588    160758   446738  717122     6829             
1247    152431    73180  708886     9994     16652   
1250    152434    74425  708886     9994     16652   
2586    160759   958863  717122     6829             

                                   authority            status  \
3353  Threatened Species Protection Act 1995        Endangered   
1251                               NT NRETAS        endangered   
904   Threatened Species Protection Act 1995        Endangered   
1249                               NT NRETAS        endangered   
932   Threatened Species Protection Act 1995        Endangered   
959   Threatened Species Protection Act 1995  Presumed extinct   
1252                               NT NRETAS        endangered   
1248                               NT NRETAS        endangered   
2588  Threatened Species Protection Act 1995        Endangered   
1247                               NT NRETAS        endangered   
1250                               NT NRETAS        endangered   
2586  Threatened Species Protection Act 1995        Endangered   

                                                    url description  \
3353  https://dpipwe.tas.gov.au/Documents/Pneumatopt...               
1251  https://lists.ala.org.au/speciesListItem/list/...               
904   https://dpipwe.tas.gov.au/Documents/Blechnum-n...               
1249  https://lists.ala.org.au/speciesListItem/list/...               
932   https://dpipwe.tas.gov.au/Documents/Blechnum%2...               
959   https://dpipwe.tas.gov.au/Documents/Botrychium...               
1252  https://lists.ala.org.au/speciesListItem/list/...               
1248  https://lists.ala.org.au/speciesListItem/list/...               
2588  https://dpipwe.tas.gov.au/Documents/Corunastyl...               
1247  https://lists.ala.org.au/speciesListItem/list/...               
1250  https://lists.ala.org.au/speciesListItem/list/...               
2586  https://dpipwe.tas.gov.au/Documents/Corunastyl...               

     geoprivacy  ...         genus specificEpithet infraspecificEpithet  \
3353   obscured  ...         Pakau       pennigera                        
1251   obscured  ...     Polytelis      alexandrae                        
904    obscured  ...        Doodia          aspera                        
1249   obscured  ...   Bellatorias           obiri                        
932    obscured  ...        Doodia         caudata                        
959    obscured  ...                                                      
1252   obscured  ...                                                      
1248   obscured  ...    Macroderma           gigas                        
2588   obscured  ...   Genoplesium        morrisii                        
1247    private  ...     Pezoporus    occidentalis                        
1250   obscured  ...  Hipposideros       inornatus                        
2586             ...   Genoplesium      nudiscapum                        

                  modified          scientificName taxonRank  \
3353  2022-09-25T23:02:37Z         Pakau pennigera   species   
1251  2019-08-27T01:09:01Z    Polytelis alexandrae   species   
904   2021-02-22T18:52:06Z           Doodia aspera   species   
1249  2019-04-30T15:19:21Z       Bellatorias obiri   species   
932   2021-02-15T17:49:20Z          Doodia caudata   species   
959                           Sceptridium australe             
1252                         Hipposideros stenotis             
1248  2019-08-27T01:58:05Z     

In [49]:
# Stats
numsensitive = len(sensitivelist.index)
numconservation = len(conservationlist.index)
numupdates  = len(updates.index)
numadditions  = len(additions.index)
numnoinatstatus = len(noinatstatus.index)
numunknownToInat = len(unknownToInat.index)
numnotaddedupdated = len(notaddedupdated.index)
numnoncomply = len(noncomply.index)
numcomply = len(statelist.index)
numdupinfo = len(dupinformation.index)
d = {'Sensitive': [numsensitive],
    'Conservation': [numconservation],
    'Statelist merge': [numfullstatelist],
    'Species iNat Comply' : [numcomply],
    'Species iNat non-Comply': [numnoncomply],
    'Duplicate Information': [numdupinfo],
    'Updates': [numupdates],
    'Additions': [numadditions],
    'Not added updated': [numnotaddedupdated],
    'No Inat Status': [numnoinatstatus],
    'Unknown to Inat': [numunknownToInat]}

statsdf = pd.DataFrame(data=d)
statsdf

Sensitive  Conservation  Statelist merge  Species iNat Comply  \
0          1           683              684                  676   

   Species iNat non-Comply  Duplicate Information  Updates  Additions  \
0                        7                      0       73        500   

   Not added updated  No Inat Status  Unknown to Inat  
0                 12             603              103